sql/size_utils.sql

```diff
@@ -368,43 +368,30 @@ $BODY$;
 -- Estimated number of rows according to catalog tables
 CREATE OR REPLACE FUNCTION approximate_row_count(relation REGCLASS)
 RETURNS BIGINT
-LANGUAGE PLPGSQL VOLATILE STRICT AS
+LANGUAGE SQL VOLATILE STRICT AS
 $BODY$
-DECLARE
-	table_name       NAME;
-	schema_name      NAME;
-	row_count_parent BIGINT;
-	row_count        BIGINT;
-BEGIN
-	SELECT relname, nspname, CASE WHEN c.reltuples > 0 THEN c.reltuples::bigint ELSE 0 END
-	INTO table_name, schema_name, row_count_parent
-	FROM pg_class c
-	INNER JOIN pg_namespace n ON (n.OID = c.relnamespace)
-	WHERE c.OID = relation;
-
-	WITH RECURSIVE inherited_id AS
-	(
-		SELECT i.inhrelid AS oid
-		FROM pg_inherits i
-		JOIN pg_class base ON i.inhparent = base.oid
-		JOIN pg_namespace base_ns ON base.relnamespace = base_ns.oid
-		WHERE base_ns.nspname = schema_name AND base.relname = table_name
-		UNION
-		SELECT i.inhrelid AS oid
-		FROM pg_inherits i
-		JOIN inherited_id b ON i.inhparent = b.oid
-	)
-	SELECT sum(CASE WHEN child.reltuples > 0 THEN child.reltuples ELSE 0 END)::bigint
-	INTO row_count
-	FROM inherited_id i
-	JOIN pg_class child ON i.oid = child.oid
-	JOIN pg_namespace child_ns ON child.relnamespace = child_ns.oid;
-
-	IF row_count IS NULL THEN
-		RETURN row_count_parent;
-	END IF;
-	RETURN row_count_parent + row_count;
-END
+  WITH RECURSIVE inherited_id AS
+  (
+    SELECT i.inhrelid AS oid
+    FROM pg_inherits i
+    WHERE i.inhparent = relation
+    UNION
+    SELECT i.inhrelid AS oid
+    FROM pg_inherits i
+    JOIN inherited_id b ON i.inhparent = b.oid
+  ),
+  parent AS
+  (
+    -- reltuples for partitioned tables is the sum of it's children in pg14 so we need to filter those out
+    SELECT CASE WHEN c.reltuples > 0 AND c.relkind <> 'p' THEN c.reltuples::bigint ELSE 0 END AS reltuples
```

Comment:
changes look fine. But won't this give incorrect counts for PG < PG14?

Reply:
No, before PG14 reltuples would just be 0 for partitioned table parent.